persistence/internal/providertest/offsetstore/transaction.go

```diff
@@ -0,0 +1,198 @@
+package offsetstore
+
+import (
+	"sync"
+
+	"github.com/dogmatiq/infix/persistence"
+	"github.com/dogmatiq/infix/persistence/internal/providertest/common"
+	"github.com/dogmatiq/infix/persistence/subsystem/offsetstore"
+	"github.com/onsi/ginkgo"
+	"github.com/onsi/ginkgo/extensions/table"
+	"github.com/onsi/gomega"
+)
+
+// DeclareTransactionTests declares a functional test-suite for a specific
+// offsetstore.Transaction implementation.
+func DeclareTransactionTests(tc *common.TestContext) {
+	ginkgo.Describe("type offsetstore.Transaction", func() {
+		var (
+			dataStore  persistence.DataStore
+			repository offsetstore.Repository
+			tearDown   func()
+		)
+
+		ginkgo.BeforeEach(func() {
+			dataStore, tearDown = tc.SetupDataStore()
+			repository = dataStore.OffsetStoreRepository()
+		})
+
+		ginkgo.AfterEach(func() {
+			tearDown()
+		})
+
+		ginkgo.Describe("func SaveOffset()", func() {
+			ginkgo.When("application has no previous offsets associated", func() {
+				ginkgo.It("saves an offset with the current offset equal to zero", func() {
+					c := offsetstore.Offset(0)
+					n := offsetstore.Offset(1)
+					saveOffset(tc.Context, dataStore, "<source-app-key>", c, n)
+
+					actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
+					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+					gomega.Expect(actual).Should(gomega.BeNumerically("==", n))
+				})
+			})
+
+			ginkgo.When("application has previous offsets associated", func() {
+				ginkgo.It("saves an offset with the current offset passed", func() {
+					c := offsetstore.Offset(0)
+					n := offsetstore.Offset(1)
+					saveOffset(tc.Context, dataStore, "<source-app-key>", c, n)
+
+					c = n
+					n += n
+
+					saveOffset(tc.Context, dataStore, "<source-app-key>", c, n)
+
+					actual, err := repository.LoadOffset(tc.Context, "<source-app-key>")
+					gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+					gomega.Expect(actual).Should(gomega.BeNumerically("==", n))
+				})
+
+				ginkgo.It("does not update the current offset with the equal next offset", func() {
```

Comment:
I'm not sure about this one. What is it guaranteeing? It's not so much that it "does not update", perhaps, but that there's no error if `c == n`?

If I have that right, it can probably be removed as "not important to the requirements of the interface", but it's not harmful either.